include/albatross/src/tune/tune.hpp

```diff
@@ -34,50 +34,194 @@ inline std::string to_string(const nlopt::result result) {
   return result_strings[result];
 }
 
-/*
- * This function API is defined by nlopt, when an optimization algorithm
- * requires the gradient nlopt expects that the grad argument gets
- * modified inside this function.  The TuneModelConfig which holds
- * any information about which functions to call etc, needs to be passed
- * in through a void pointer.
- */
-template <typename ModelType, typename MetricType, typename FeatureType>
-double objective_function(const std::vector<double> &x,
-                          std::vector<double> &grad, void *void_tune_config) {
-  if (!grad.empty()) {
-    throw std::invalid_argument("The algorithm being used by nlopt requires"
-                                "a gradient but one isn't available.");
+template <typename ObjectiveFunction>
+inline double objective_function_wrapper(const std::vector<double> &x,
+                                         std::vector<double> &grad,
+                                         void *objective_func) {
+  return (*static_cast<ObjectiveFunction *>(objective_func))(x, grad);
+}
+
+template <typename ObjectiveFunction>
+inline void set_objective_function(nlopt::opt &optimizer,
+                                   ObjectiveFunction &objective) {
+  optimizer.set_min_objective(objective_function_wrapper<ObjectiveFunction>,
+                              (void *)&objective);
+}
+
+inline nlopt::opt
+default_optimizer(const ParameterStore &params,
+                  const nlopt::algorithm &algorithm = nlopt::LN_SBPLX) {
+  // The various algorithms in nlopt are coded by the first two characters.
+  // In this case LN stands for local, gradient free.
+  const auto tunable_params = get_tunable_parameters(params);
+
+  nlopt::opt optimizer(algorithm, (unsigned)tunable_params.values.size());
+  optimizer.set_ftol_abs(1e-8);
+  optimizer.set_ftol_rel(1e-6);
+  optimizer.set_lower_bounds(tunable_params.lower_bounds);
+  optimizer.set_upper_bounds(tunable_params.upper_bounds);
+  // the sensitivity to parameters varies greatly between parameters so
+  // terminating based on change in x isn't a great criteria, we only
+  // terminate based on xtol if the change is super small.
+  optimizer.set_xtol_abs(1e-18);
+  optimizer.set_xtol_rel(1e-18);
+  return optimizer;
+}
+
+inline ParameterStore uniformative_params(const std::vector<double> &values) {
```

Comment:
`uniformative` should be a word meaning it forms one thing...

Reply:
Oppps, typo it should be `uninformative`